pkg/virt-controller/watch/snapshot/snapshot.go

```diff
@@ -686,24 +688,66 @@ func (ctrl *VMSnapshotController) getVolumeStorageClassForVolume(namespace strin
 		return ctrl.getVolumeSnapshotClass(*pvc.Spec.StorageClassName)
 	}
 	if volume.VolumeSource.DataVolume != nil {
-		pvcKey := cacheKeyFunc(namespace, volume.VolumeSource.DataVolume.Name)
-		// TODO Change when PVC rename PR is implemented
-
-		obj, exists, err := ctrl.PVCInformer.GetStore().GetByKey(pvcKey)
+		storageClassName, err := ctrl.getStorageClassNameForDV(namespace, volume.VolumeSource.DataVolume.Name)
 		if err != nil {
 			return "", err
 		}
 
-		if !exists {
-			log.Log.V(3).Infof("PVC not in cache [%s]", pvcKey)
-			return "", fmt.Errorf("PVC for the DataVolume not found")
+		return ctrl.getVolumeSnapshotClass(storageClassName)
+	}
+
+	return "", fmt.Errorf("Volume type does not suport snapshots")
+}
+
+func (ctrl *VMSnapshotController) getStorageClassNameForDV(namespace string, dvName string) (string, error) {
+	// First, look up DV's StorageClass
+	key := cacheKeyFunc(namespace, dvName)
+
+	obj, exists, err := ctrl.DVInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return "", err
+	}
+
+	var dv *cdiv1.DataVolume
+	if exists {
+		dv = obj.(*cdiv1.DataVolume)
+		if dv.Spec.PVC == nil || dv.Spec.PVC.StorageClassName == nil {
+			log.Log.V(3).Infof("DV [%s] not found", key)
+			return "", fmt.Errorf("DV '%s' not found", key)
 		}
+		if *dv.Spec.PVC.StorageClassName != "" {
+			return *dv.Spec.PVC.StorageClassName, nil
+		}
+	} else {
+		log.Log.V(3).Infof("DV not in cache [%s]", key)
+		return "", nil
+	}
 
-		pvc := obj.(*corev1.PersistentVolumeClaim)
-		return ctrl.getVolumeSnapshotClass(*pvc.Spec.StorageClassName)
+	// Second, see if DV is owned by a VM, and if so, if the DVTemplate has a StorageClass
+	for _, or := range dv.OwnerReferences {
+		if or.Kind == "VirtualMachine" {
+			log.Log.V(3).Infof("Checking VM %s", or.Name)
+
+			vmKey := cacheKeyFunc(namespace, or.Name)
+			storeObj, exists, err := ctrl.VMInformer.GetStore().GetByKey(vmKey)
+			if err != nil || !exists {
+				continue
+			}
+
+			vm, ok := storeObj.(*kubevirtv1.VirtualMachine)
+			if !ok {
+				continue
+			}
+
+			for _, dvTemplate := range vm.Spec.DataVolumeTemplates {
```

Comment:
Could you explain why one has to look up a VM? If there is no real DV, why do I have to look up a storage class for it and not just do nothing for the DV?

Reply:
Could wait for the DV to be created but why wait if the info is in the template?

Reply:
This point is reached not only when the DV is not found, but also when the StorageClass for the DV has not been found. Previously we waited for the PVC to be ready to read its StorageClass, but this has a side effect: the status is (possibly) wrongly set to `false` and later updated to a whatever it should be (true or false).

Reply:
Let me rephrase my question: If the DV and PVC does not exist, can we ever reach this point?

Reply:
No, we can't.

Let me refactor the code to make it clearer. Also, I seem to have lost the "last resort" case of looking at the PVC's StorageClass.